### PR TITLE
python@3.10: migrate aliases

### PR DESCRIPTION
**Pull request by XuehaiPan**
- [X] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [X] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [X] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [X] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

Migrate aliases (`python`, `python3`, `python@3`) to `python@3.10`.

### CODE REVIEW

**Comment by carlocab** [comment]
Not sure this is ready yet, as `python@3.10` is still keg-only. Open to other opinions, however.

**Comment by XuehaiPan** [comment]
> The migration for `python@3.9` took about 3 weeks.

The migration process for `python@3.10` takes longer than expected since we have more formulae than a year ago. Be patient.

```console
$ git rev-parse HEAD  # current HEAD
ec64d6dd8537b662ecb84a500508a89aa12e77fe

$ for formula in $(find Formula -name 'python@*.rb' | xargs basename -s .rb | sort -rV); do echo "$(grep -rF 'depends_on "'"${formula}"'"' Formula | wc -l)\tformulae depends on ${formula}"; done
268     formulae depends on python@3.10
335     formulae depends on python@3.9
11      formulae depends on python@3.8
3       formulae depends on python@3.7

$ git checkout a4d29eb6431ac9b57f07d6b29f3c5a1238ef2370
Note: switching to 'a4d29eb6431ac9b57f07d6b29f3c5a1238ef2370'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:

  git switch -c <new-branch-name>

Or undo this operation with:

  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

HEAD is now at a4d29eb6431 python@3.9: migrate aliases

$ for formula in $(find Formula -name 'python@*.rb' | xargs basename -s .rb | sort -rV); do echo "$(grep -rF 'depends_on "'"${formula}"'"' Formula | wc -l)\tformulae depends on ${formula}"; done
425     formulae depends on python@3.9
22      formulae depends on python@3.8
5       formulae depends on python@3.7
```

**Comment by carlocab** [comment]
I wouldn't count on the remaining migration being completed in three weeks. It might, but I also think it'll take longer than the first batch of formulae that's been migrated, since most of the remaining formulae to be migrated live in a few large dependency trees that need to be migrated together. That will take some work.

**Comment by carlocab** [comment]
```
❯ git rev-parse HEAD
252d93c00df9a7e2d809c9fbd723de7334dc8379

❯ for formula in $(find Formula -name 'python@*.rb' | xargs basename -s .rb | sort -rV); do echo "$(grep -rF 'depends_on "'"${formula}"'"' Formula | wc -l)\tformulae depend on ${formula}"; done
274     formulae depend on python@3.10
331     formulae depend on python@3.9
11      formulae depend on python@3.8
3       formulae depend on python@3.7
```
Not that much movement in the past ten days.

**Comment by XuehaiPan** [comment]
Any updates? It has been two months since Python 3.10 was released.

**Comment by SMillerDev** [comment]
Running the script now gives 10 more formula on 3.10. You can't just wait here, if you want this to happen soon you'll need to help migrate them.
```
     284\tformulae depend on python@3.10
     325\tformulae depend on python@3.9
      11\tformulae depend on python@3.8
       3\tformulae depend on python@3.7
```

**Comment by cclauss** [comment]
Please add [label: python-3.10-migration](https://github.com/Homebrew/homebrew-core/labels/python-3.10-migration) so that those who want to help get to Python 3.10 know how to discover which formulae to work on.

**Comment by XuehaiPan** [comment]
I think we could open an issue to track the progress of the Python 3.10 migration process.

Should we migrate formula-by-formula in a single pull request or migrate bunch of formulae in a single pull request (test them at once)?

**Comment by carlocab** [comment]
It's not quite that simple, since dependency trees need to be migrated together. See https://github.com/Homebrew/homebrew-core/pull/87277

**Comment by XuehaiPan** [comment]
> It's not quite that simple, since dependency trees need to be migrated together. See #87277

There are 6 pull requests needed in total:

### 1. `gdb` family

```
[['gdb'], ['gdbgui']]
```

### 2. `llvm@11` family

```
[['llvm@11'], ['chapel']]
```

### 3. `mercurial` family

```
[['mercurial'], ['hg-fast-export']]
```

### 4 `molten-vk` family

```
[['molten-vk'], ['ppsspp']]
```

### 5. `node@14` family

```
[['node@14'], ['code-server']]
```

### 6. `six`, `rust`, `glib` family

```
[
  [
    "six",
    "rust",
    "glib",
    "python-tabulate",
    "llvm",
    "cython",
    "pillow",
    "libxml2",
    "py3cairo",
    "sip",
    "pybind11",
    "tbb",
    "llvm@12",
    "jsoncpp",
    "libxcb",
    "jack",
    "gpgme",
    "opencolorio",
    "volk",
    "libepoxy",
    "talloc",
    "xapian",
    "cryptominisat",
    "uhd"
  ],
  [
    "numpy",
    "gobject-introspection",
    "protobuf",
    "itstool",
    "node",
    "pyqt-builder",
    "mesa",
    "fontforge",
    "vapoursynth",
    "stp",
    "wllvm",
    "mpi4py",
    "notmuch",
    "s3cmd",
    "ncspot",
    "gtk-doc",
    "checkov",
    "aws-google-auth",
    "ciphey",
    "sceptre",
    "fabric",
    "regipy",
    "principalmapper",
    "redex",
    "deno",
    "ykman",
    "breezy",
    "q",
    "mycli",
    "magic-wormhole",
    "pgcli",
    "cassandra",
    "lcm",
    "osinfo-db-tools",
    "credstash",
    "snapcraft",
    "libvirt",
    "homeassistant-cli",
    "px",
    "moto",
    "pgxnclient",
    "vsts-cli",
    "pywhat",
    "v8",
    "ansible@2.9",
    "ghc",
    "ispc",
    "termius",
    "charm-tools",
    "lasso",
    "theharvester",
    "lensfun",
    "ocrmypdf",
    "woob",
    "sqlite-utils",
    "jrnl",
    "athenacli",
    "c7n",
    "sslyze",
    "pyflow",
    "awscli@1",
    "localstack",
    "dxpy",
    "profanity",
    "awscli",
    "halide",
    "charmcraft",
    "buku",
    "cruft",
    "proselint",
    "jenkins-job-builder",
    "libsigrokdecode",
    "pwntools",
    "flintrock",
    "remarshal",
    "platformio",
    "fava",
    "py-spy",
    "howdoi",
    "pocsuite3",
    "todoman",
    "trezor-agent",
    "amp",
    "robot-framework",
    "duplicity",
    "lit",
    "suricata",
    "litecli"
  ],
  [
    "pygobject3",
    "boost-python3",
    "at-spi2-core",
    "pythran",
    "libtorch",
    "apache-arrow",
    "texlive",
    "gdal",
    "opencv@3",
    "mavsdk",
    "mpv",
    "spice-gtk",
    "csound",
    "yelp-tools",
    "aubio",
    "nest",
    "anime-downloader",
    "mftrace",
    "ola",
    "easy-tag",
    "evince",
    "llnode",
    "gupnp",
    "libtensorflow",
    "klee",
    "anjuta",
    "polynote",
    "mitmproxy",
    "jupyterlab",
    "fdroidserver",
    "root",
    "adios2",
    "alot",
    "emscripten",
    "gst-devtools",
    "wxpython",
    "astrometry-net",
    "arcade-learning-environment",
    "binwalk",
    "gucharmap",
    "bzt",
    "nanopb-generator",
    "libgusb",
    "gom"
  ],
  [
    "qt@5",
    "qt",
    "scipy",
    "torchvision",
    "openimageio",
    "libtorrent-rasterbar",
    "pympress",
    "xdot",
    "gst-python",
    "terminator",
    "lilypond",
    "pytouhou",
    "dvc",
    "at-spi2-atk",
    "gexiv2",
    "libpeas",
    "zim",
    "mapserver",
    "mat2",
    "libsigrok",
    "rdkit",
    "eigenpy",
    "nicotine-plus"
  ],
  [
    "pyqt@5",
    "pyqt",
    "pyside@2",
    "pyside",
    "graph-tool",
    "mame",
    "networkit",
    "pushpin",
    "rom-tools"
  ],
  [
    "vtk@8.2",
    "vtk",
    "carla",
    "pyqt-3d",
    "pyqt-networkauth",
    "qscintilla2",
    "trimage",
    "coin3d",
    "git-cola",
    "gnuradio",
    "urh"
  ],
  [
    "opencv",
    "gdcm"
  ],
  [
    "openalpr"
  ]
]
```

**Comment by cclauss** [comment]
Instead of `json.dump()`, could the code dump the results as markdown todo list:
```markdown
### stand-alone formulae (111)
- [ ] arangodb
- [ ] auditbeat
- [ ] bandcamp-dl
- [ ] bedtools
- [x] binaryen
- [ ] blockhash
...
```
So that contributors have a todo list to work on:
### 111 stand-alone formulae
- [ ] arangodb
- [ ] auditbeat
- [ ] bandcamp-dl
- [ ] bedtools
- [x] binaryen
- [ ] blockhash
...

**Comment by carlocab** [comment]
You can remove `six`, `pillow`, and `protobuf` -- those do not need to be migrated.

**Comment by cclauss** [comment]
Well done @XuehaiPan 
```
     336	formulae depend on python@3.10
     274	formulae depend on python@3.9
      11	formulae depend on python@3.8
       3	formulae depend on python@3.7
```

**Comment by XuehaiPan** [comment]
Can we treat those `depends_on "python@3.9" => :build` formulae (in https://github.com/Homebrew/homebrew-core/pull/87075#issuecomment-988102123 Family `glib` 202) as stand-alones? I'm not so sure.

**Comment by cclauss** [comment]
Please resolve the git conflicts on the PR.

**Comment by XuehaiPan** [comment]
> Please resolve the git conflicts on the PR.

I think is good to block the CI tests with these conflicts, since we are not ready to merge.

**Comment by cclauss** [comment]
The criterion above https://github.com/Homebrew/homebrew-core/pull/87075#issuecomment-945629226 was:
> We should wait until most/all Python-dependent formulae have been migrated to python@3.10 first

I believe that we have ___most___ and that ___all___ is not reasonable (see TensorFlow, etc.) so what should be the target?

**Comment by XuehaiPan** [comment]
> I believe that we have **most** and that **all** is not reasonable (see TensorFlow, etc.) so what should be the target?

In some formulae's `Makefile`, they use option `PYTHON ?= python3`. If the formulae are added when `python@3.9` is the default `python3`, the contributor can have the following with no test error:

```ruby
depends_on "python@3.9"

def install
    system "make"
end

def test
    ...
end
```

When migrating to `python@3.10`, there is no `python3` executable in `PATH` in the test phase. Then the tests will fail. For example:

- PR: #91019
- CI test: https://github.com/Homebrew/homebrew-core/runs/4491819335

They should have:

```ruby
depends_on "python@3.9"

def install
    system "make", "PYTHON=#{which("python3")}"  # there is `opt/python@3.9/bin/python3`
end

def test
    ...
end
```

Ideally, when a new formula is added (including build-onlies and test-onlies), they should try to test with the key-only Python as well.

**Comment by cclauss** [comment]
58.81% of these formulae depend on python@3.10
```
     367	formulae depend on python@3.10
     243	formulae depend on python@3.9
      11	formulae depend on python@3.8
       3	formulae depend on python@3.7

python3 -c "print( f'{367 / (367+243+11+3) * 100:.2f}% of these formulae depend on python@3.10')"
```

**Comment by cclauss** [comment]
Which formulas should be worked on next?

**Comment by cclauss** [comment]
__60.35%__ of these formulae depend on python@3.10

`for formula in $(find Formula -name 'python@*.rb' | xargs basename -s .rb | sort -rV); do echo "$(grep -rF 'depends_on "'"${formula}"'"' Formula | wc -l)\tformulae depend on ${formula}"; done`
```
     379	formulae depend on python@3.10
     235	formulae depend on python@3.9
      11	formulae depend on python@3.8
       3	formulae depend on python@3.7
```
`python3 -c "print( f'{379 / (379+235+11+3) * 100:.2f}% of these formulae depend on python@3.10')"`

**Comment by cclauss** [comment]
__62.03%__ of these formulae depend on python@3.10
```
     392	formulae depend on python@3.10
     226	formulae depend on python@3.9
      11	formulae depend on python@3.8
       3	formulae depend on python@3.7
```
`python3 -c "print( f'{392 / (392+226+11+3) * 100:.2f}% of these formulae depend on python@3.10')"`

@XuehaiPan Any advice on how to discover which formulae to upgrade next?

**Comment by XuehaiPan** [comment]
> Any advice on how to discover which formulae to upgrade next?

Maybe the formulae with the build-only and test-only dependency of `python@3.9`.

The migration of remaining formulas (not build/test-only) may need to be done in a single PR by Homebrew members, which requires extensive CI testing.

**Comment by carlocab** [comment]
Since we're already modifying `python@3.9` here, I think we should consider doing `make altinstall` with it, so that it doesn't need to be keg-only. This makes it so that it can be installed alongside a different version of Python without conflicts.

See https://github.com/python/cpython#installing-multiple-versions.

**Comment by cclauss** [comment]
__64.55%__ of these formulae depend on python@3.10

See [calculation above](https://github.com/Homebrew/homebrew-core/pull/87075#issuecomment-999541562).

**Comment by cclauss** [comment]
Sorry for the interruption here but...

By everyone's accounts Go v1.18 release was significant, monumental, and the biggest release yet with the release of generics, fuzzing, etc.  Nevertheless, v1.18 became the Homebrew aliased release of Go within **5 days** of its release.  Admittedly, there was lots of prep and community collaboration to make that happen.  #91369 

Python v3.10 was not a significant release yet it is still not the Homebrew aliased release of Python even **5 months** after its ship date.  Python 3.10.3 is now the current release yet Homebrew's default install is still Python 3.9. There have been scads of PRs landed so that now a supermajority of Homebrew Python packages rely on Py3.10.  The remaining packages must be migrated in a single PR by someone on the Homebrew core team https://github.com/Homebrew/homebrew-core/pull/87075#issuecomment-1017585018

I have asked several times above (and elsewhere) how we can reestablish momentum on this PR.  How can I and others help?

Given that Python 3.11 is in its final alpha releases, it would be useful if we could find a way to push this PR forward and figure out how to accelerate future Python release processes.

**Comment by iMichka** [comment]
> Admittedly, there was lots of prep and community collaboration to make that happen. https://github.com/Homebrew/homebrew-core/pull/91369

Actually for `go` a RC1 PR was opened, 122 packages were broken, and nobody fixed them. The preparation was useless and there was no community collaboration (just a lot of complaints, and pressure). Finally one maintainer, @Bo98, had to step in an fixed all the broken things by himself, with the help of a few other maintainers who reviewed everything.

Regarding the momentum here: these things take time, and motivation comes and goes. The good thing is that there was more community help for `Python` than for `go`.

So yeah, right now it boils down to some maintainers stepping up and finding the energy to migrate the remaining formulae. As it needs a lot of CI, we need to squeeze these builds in our already overbooked build schedule. So yeah, Python is on our radar. And we are working on it, but slowly.

File: Formula/python@3.10.rb [hunk 1]
```diff
@@ -361,11 +380,10 @@ def sitecustomize
           # This can only happen if the user has set the PYTHONPATH to a mismatching site-packages directory.
           # Every Python looks at the PYTHONPATH variable and we can't fix it here in sitecustomize.py,
           # because the PYTHONPATH is evaluated after the sitecustomize.py. Many modules (e.g. PyQt4) are
-          # built only for a specific version of Python and will fail with cryptic error messages.
```
Comment (stephenwade) [inline Formula/python@3.10.rb:364]: I think this line shouldn't be removed.

**Comment by fanninpm** [comment]
@XuehaiPan here are some formulae that previously had CI errors but now have successfully migrated to `python@3.10`:

- [`bandcamp-dl`](https://github.com/Homebrew/homebrew-core/pull/91874)
- [`shallow-backup`](https://github.com/Homebrew/homebrew-core/pull/91091)
- [`spades`](https://github.com/Homebrew/homebrew-core/pull/94783)

**Comment by cclauss** [comment]
Any advice on how we repair this in time for [Python 3.11](https://www.python.org/downloads/release/python-3110b1)?

**Comment by alexreg** [comment]
Already there are more formula that depend on Python 3.10 than 3.9. I'm not sure what we're waiting for now.

**Comment by XuehaiPan** [comment]
> Already there are more formulas that depend on Python 3.10 than 3.9. I'm not sure what we're waiting for now (to make the switch with `python -> python@3.10`).

There still are lots of formulae depending on `glib`. And `glib` depends on `python@3.9`. All these formulae need to be migrated in a single PR.

Suppose there is a formula that depends on `glib` but the upstream does not support Python 3.10 yet. Such as:

```diff
-from collections import Mapping  # removed in Python 3.10
+from collections.abc import Mapping
```

It will block our migration process.

**Comment by alexreg** [comment]
@XuehaiPan 

I see. Well, that's sort of fair enough.

On the other hand, given glib is by far a native library, surely it would be easy to patch the small bits of Python to support 3.10? Just a thought.

**Comment by iMichka** [comment]
I think it's time we split `glib` in two separate packages. Conda did it too: https://uwekorn.com/2020/10/19/r-without-python.html

**Comment by alexreg** [comment]
> I think it's time we split `glib` in two separate packages. Conda did it too: [uwekorn.com/2020/10/19/r-without-python.html](https://uwekorn.com/2020/10/19/r-without-python.html)

I agree, this might be handy. Let's see what maintainers think.

**Comment by davecom** [comment]
This might be a dumb question—but I am sincerely curious. Why is it necessary that every package be migrated to the latest version? Why can't the generic python package evolve almost immediately to the latest version of Python, and individual dependent packages simply be versioned to any particular python and migrate to a newer version at their own pace?

Thank you to all of you who work on the Python Homebrew ecosystem.

**Comment by SMillerDev** [comment]
> This might be a dumb question—but I am sincerely curious. Why is it necessary that every package be migrated to the latest version? 

Because that's the way Homebrew works. It uses the latest possible dependencies of everything. It avoids maintainers writing patches for old software because the new versions are usually developed against the latest dependencies.

**Comment by xxyzz** [comment]
> What needs to be ported in `glib`? I was unable to find the following code in the `glib` repo.
> 
> ``` python
> from collections import Mapping
> ```

I think he means this line of code is used in a Python formula which depends on `glib`.

BTW, I find Arch Linux marks `python` as optional dependency for some `glib2` commands(gdbus-codegen, glib-genmarshal, glib-mkenums, gtester-report). Does this means if packages that depend on glib are compatible with Python 3.10 or don't depends on Python, update Python shouldn't affect them?

**Comment by XuehaiPan** [comment]
Can we add a virtual formula `python@any`? Then log the actual Python formula in `install` to the `opt/<formula>/.brew` folder.

`glib` does not require a specific version of Python (any `>=3.5`) and does not install any package in `site-packages`.

**Comment by XuehaiPan** [comment]
I create a discussion (<https://github.com/orgs/Homebrew/discussions/3274>) for a new minimal runtime-only Python formula.

**Comment by xxyzz** [comment]
Only 11 packages depends on `glib` and `python` directly(`brew uses glib python`):
- fontforge
- lcm
- libsigrok 
- osinfo-db-tools
- gobject-introspection  
- lensfun     
- libsigrokdecode         
- qt
- gupnp                   
- libpeas              
- notmuch

**Comment by SMillerDev** [comment]
> and also that there is no reason to block on this upgrade, given the small number of glib + python formulas.

A small number of formula that are broken are still a number of formula that are broken and should therefore be avoided.

**Comment by XuehaiPan** [comment]
> Only 10 packages depend on both `glib` and `python` directly (`brew uses glib python`)

Migrate `glib` to `python@3.10` may break formulae that have a deeper level of dependency trees.

```console
$ brew uses --recursive glib python@3.9 | wc -l                       
687

$ brew uses --recursive glib python@3.10 | wc -l
340

$ for formula in $(find Formula -name 'python@*.rb' | xargs basename -s .rb | sort -rV); do echo "$(grep -rlF 'depends_on "'"${formula}"'"' Formula | wc -l)\tformulae depend on ${formula}"; done
500	formulae depend on python@3.10
168	formulae depend on python@3.9
10	formulae depend on python@3.8
3	formulae depend on python@3.7
```

**Comment by MikeMcQuaid** [comment]
> > This might be a dumb question—but I am sincerely curious. Why is it necessary that every package be migrated to the latest version?
> 
> Because that's the way Homebrew works. It uses the latest possible dependencies of everything. It avoids maintainers writing patches for old software because the new versions are usually developed against the latest dependencies.

Also, at some point the older version will need to be removed because it's broken and/or upstream are no longer providing security updates for it. If we don't migrate packages before then, we end up having to do so against the clock of potentially vulnerable software for our users.

**Comment by SMillerDev** [comment]
> This stance about always requiring packages to be using dependencies that are the latest is leading most Homebrew users to be using an older version of Python 7 months after the arrival of a major point release.

Users that do not configure their system for a specific python version*. Users that wanted a specific version of python could have had it on the 3rd of September by adjusting their path to include it.

The main issue here is mostly the same as with Go updates. It's a lot of work and Homebrew volunteers have limited time and motivation available to deal with it.

**Comment by davecom** [comment]
> > This stance about always requiring packages to be using dependencies that are the latest is leading most Homebrew users to be using an older version of Python 7 months after the arrival of a major point release.
> 
> Users that do not configure their system for a specific python version*. Users that wanted a specific version of python could have had it on the 3rd of September by adjusting their path to include it.

I'm trying to think through the lens of the average user. I teach CS—I know the person just learning Python for example is not going to know how to do that, so they're just going to use whatever is provided and wonder occasionally why the examples on Python.org don't match up with what they can do.

> 
> The main issue here is mostly the same as with Go updates. It's a lot of work and Homebrew volunteers have limited time and motivation available to deal with it.

Totally sympathize and like I said before, I'm not the one doing the hard work. That said, a policy of allowing dependent packages to float for up to N releases back would actually decrease the time pressure on the volunteers who maintain Homebrew. It would give them years to move packages forward instead of being pressured by users like those on this thread to do it quickly.

**Comment by cclauss** [comment]
Any news on this effort?  It has been more than 250 days since Python 3.10 shipped yet it is still not Homebrew's default.  

Python 3.11 beta 3 is now out and [when it is released](https://peps.python.org/pep-0664/#release-schedule) 3.11 will include [a 10-60% speedup](https://docs.python.org/3.11/whatsnew/3.11.html#summary-release-highlights) and [support for cross-compiling to WebAssembly](https://github.com/python/cpython/issues/84461) so a lot of Pythonista's will be interested in making that upgrade.  It could save a ton of CPU hours if Homebrew could quickly make that upgrade its default Python.

Is there anything someone not on the Homebrew core team can do to help establish a smoother transition to making the current Python version the default on Homebrew?

**Comment by fanninpm** [comment]
@cclauss  can you please consider taking over @XuehaiPan's efforts to split up `glib` into two separate formulae? Complaining about the state of things is unlikely to yield results. Otherwise, if you want Python 3.10 to be the default at this point, please consider using a different package manager, such as [Nix](https://github.com/NixOS/nixpkgs).

**Comment by SMillerDev** [comment]
> Is there anything someone not on the Homebrew core team can do to help establish a smoother transition to making the current Python version the default on Homebrew?

As far as I know the work that needs to be done has been laid out in this issue. You don't need to be in the homebrew team to pick those things up.

**Comment by danielnachun** [comment]
I ran into this while working the `qt@5` upgrade and it's very clear that few or no formulae depend on the `glib` Python scripts - it doesn't provide any Python wheels.  For this reason I am heavily leaning towards using the conflicting versioned dependencies allowlist to ignore any conflicts from `glib`, which is almost exclusively the cause of the current blockage.  

When https://github.com/Homebrew/homebrew-core/pull/103229 is merged, I am going to open a PR to migrate `pyqt@5`, `py3cairo`, `pygobject3` their dependents to `python@3.10` because they do have to be migrated together.  I believe this is the largest "real" dependency tree that needs to moved like this, though if I identify any other trees I'll open PRs for those.  Most of the remaining formulae are really standalone formulae that only pick up a `python@3.9` dependency through `glib` (or less often `tbb`, which is also probably spurious).

Once the `glib` formula split is finished, we will be able to remove most of the formulae from the conflicting versioned dependencies allowlist and can switch the default to `python@3.10`.  More importantly, the `glib` split should make the migration to `python@3.11` much less difficult.

**Comment by iMichka** [comment]
The `glib` split effort is now taken care of here: #103916
I am also working on the `python-tabulate` split, which should also help: #99064

**Comment by danielnachun** [comment]
Looking more carefully at our current situation, I think we should also made the `pyqt` and `pyside` formulae "multi-Python" as well.  This will enable us to migrate their dependents individually.  In fact we may want to make it mandatory that Python formulae which install wheels used by other packages always be built this way when a new Python comes out.  This would allow us to migrate almost all formulae individually instead of relying on big PRs with multiple formulae that run into bottlenecks we've currently been struggling with.

**Review by MikeMcQuaid — DISMISSED** [review]

**Comment by danielnachun** [comment]
There have been some updates on this.  I have made the `pyqt@5`, `py3cairo` and `pygobject3` formulae into multi-Python packages that support both Python 3.9 and 3.10, and we changed `tbb` to only have a build dependency on `python@3.10` so that will no longer cause conflicts either.

The remaining issues are: 
1) Merge https://github.com/Homebrew/homebrew-core/pull/103916 to split out `glib` and `glib-utils`.
2) See if we can use a similar solution for `gobject-introspection`.  Like `glib`, this formula has some Python executables but ships no Python libraries.
3) (Optional) Split out `lldb` from `llvm`: https://github.com/Homebrew/homebrew-core/pull/103879.  There are not many `llvm` dependents on macOS so this isn't a major blocker.  We can just use the allowlist for this if this ends up taking a while.

**Comment by XuehaiPan** [comment]
CI tap syntax has passed. No more conflict in the dependency trees.

```console
$ git rev-parse HEAD
40171bc81d27ffbd0c53e5503f57365c4d901670

$ for formula in $(find Formula -name 'python@*.rb' | xargs basename -s .rb | sort -rV); do echo "$(grep -rlF 'depends_on "'"${formula}"'"' Formula | wc -l)\tformulae depend on ${formula}"; done
599     formulae depend on python@3.10
103     formulae depend on python@3.9
9       formulae depend on python@3.8
2       formulae depend on python@3.7
```

**Comment by carlocab** [comment]
> Since we're already modifying `python@3.9` here, I think we should consider doing `make altinstall` with it, so that it doesn't need to be keg-only. This makes it so that it can be installed alongside a different version of Python without conflicts.
> 
> See https://github.com/python/cpython#installing-multiple-versions.

Implemented this at #107517.

**Comment by XuehaiPan** [comment]
> Since we're already modifying python@3.9 here, I think we should consider doing `make altinstall` with it, so that it doesn't need to be keg-only. This makes it so that it can be installed alongside a different version of Python without conflicts.

Since `make altinstall` does not make a symlink `python3` to `python3.x` in the prefix, it will breaks dependent formulae that use:

```ruby
depends_on "python@3.9"

python = Formula["python@3.9"].bin/"python3"
```

or 

```ruby
depends_on "python@3.9"

python = which("python3")
```

There are bunch of formulae need to add `.9` suffix as `Formula["python@3.9"].bin/"python3.9"`. All of the dependent update should be included in PR #107517. It would be a huge PR with lots of file changes.

If we make a symlink `prefix/bin/python3` to `prefix/bin/python3.x`, it will break `brew link`.

**Comment by carlocab** [comment]
For those formulae in particular:

There aren't too many of those. I count less than 35. But it is not necessary to fix them all in the same PR.

Some alternatives:
- fix them to look for `python3.9` instead of `python3` separately.
- add `python@3.9`'s `libexec/"bin"` to `PATH` during the build. This can even be handled in `brew`.

The second doesn't completely fix the ones that look for `python3` at a specific path, but that can be adjusted in a separate PR too.

Of course, there could be many more formulae that assume `python@3.9`'s `python3` is in `bin`, but I'm not too concerned about fixing them all, especially if this is a short-term transitional cost we pay in order to make future Python migrations easier.

**Comment by XuehaiPan** [comment]
Require 91 + 38 + 28 separate PRs to add `major_minor` to the python executable suffix.

```console
$ for formula in $(find Formula -name 'python@*.rb' | xargs basename -s .rb | sort -rV); do echo "$(grep -rlF 'Formula["'"${formula}"'"].opt_bin' Formula | wc -l)\tformulae use python3 in ${formula}'s opt_bin"; done
91      formulae use python3 in python@3.10's opt_bin
38      formulae use python3 in python@3.9's opt_bin
0       formulae use python3 in python@3.8's opt_bin
0       formulae use python3 in python@3.7's opt_bin

$ grep -rlF 'which("python3")' Formula | wc -l
28
```

It would be good to handle this in `brew`.

We can also add new attributes to `python@3.x` formulae:

```ruby
def python3
    bin/"python#{version.major_minor}"
end

def python
    bin/"python#{version.major_minor}"
end
```

Then in the dependent:

```diff
-python = Formula["python@3.9"].bin/"python3"
+python = Formula["python@3.9"].python3
```

File: .github/workflows/triage.yml [hunk 2]
```diff
@@ -162,7 +162,7 @@ jobs:
               keep_if_no_match: true
 
             - label: CI-linux-self-hosted
-              path: Formula/(gdbm|go|openssl@1.1|python@3.9).rb
+              path: Formula/(gdbm|go|openssl@1.1|python@3.10).rb
```
Comment (cho-m) [inline .github/workflows/triage.yml:165]: Can you create a separate PR for this? Previous PR runs show that `python@3.10` needs `CI-linux-self-hosted`. Last `python@3.9` run (#107088) took less than 2 hours, so most likely doesn't need self-hosted anymore.

**Comment by cclauss** [comment]
Python 3.11 release candidate 1 has just shipped at https://www.python.org/download/pre-releases

It should accelerate programs by ~22% over previous versions of Python when released (scheduled for 2022-10-03).

**Comment by carlocab** [comment]
Thanks for your help with pushing the migration, @XuehaiPan.

**Pull request closed without merging**